### DISCUSSION TITLE
TitlePart extension

### DISCUSSION THREAD

**Discussion by OlziiNaran** [OrchardCMS/OrchardCore#9476] [Q&A] [answered]
I want to extend TitlePart and add a language-based option to its settings. When a language is selected, the settings will be displayed and effective. How should I expand?

**Comment by EricBui619** [comment, marked as answer]
@OlziiNaran As I understand, you wanna make something like multi-language, is it right? So before you confirm my question, I can give you some suggestions. About TitlePart, it's just a class that is inherited from ContentPart class of OC and only have one field named Title, so  it's easy for you to create new TitlePart replacing or inheriting the old one. Your new TitilePart can have any additional field as you want.

One more important thing, let's go into OC source code, search TitilePartHandler. In this class, you will find some methods to handle CRUD operations of content part. Also, It has a private method called GetSettings that OC use to get settings of title part. Let's take a look at it, that will give you some ideas before we continue to do other steps.

![image](https://user-images.githubusercontent.com/30250345/118920267-4d4fb380-b960-11eb-8262-1bd36b72d5af.png)

![image](https://user-images.githubusercontent.com/30250345/118920294-5e98c000-b960-11eb-9d88-b7ca89a73d5a.png)

**Reply by OlziiNaran** [reply]
Based Part development and core upgrade, which is better?

**Reply by EricBui619** [reply]
Maybe Based Part development, core upgrade if you understand everything clearly and certainly

**Reply by OlziiNaran** [reply]
> @OlziiNaran As I understand, you wanna make something like multi-language, is it right? So before you confirm my question, I can give you some suggestions. About TitlePart, it's just a class that is inherited from ContentPart class of OC and only have one field named Title, so it's easy for you to create new TitlePart replacing or inheriting the old one. Your new TitilePart can have any additional field as you want.
> 
> One more important thing, let's go into OC source code, search TitilePartHandler. In this class, you will find some methods to handle CRUD operations of content part. Also, It has a private method called GetSettings that OC use to get settings of title part. Let's take a look at it, that will give you some ideas before we continue to do other steps.
> 
> ![image](https://user-images.githubusercontent.com/30250345/118920267-4d4fb380-b960-11eb-8262-1bd36b72d5af.png)
> 
> ![image](https://user-images.githubusercontent.com/30250345/118920294-5e98c000-b960-11eb-9d88-b7ca89a73d5a.png)

Thank you for your suggestion. According to your explanation, my idea is basically realized

**Comment by Skrypt** [comment]
Normally, you should just add the LocalizationPart to your Content Type so that you can translate it's entire content. Which removes the need to extend the TitlePart specifically. But, if you want to be able to localize only the TitlePart while keeping other content to be non-translated then I do not recommend this approach.

**Reply by OlziiNaran** [reply]
Thanks for the reply, as you said, I need to change some part or field configuration items based on language settings

**Reply by jptissot** [reply]
I don't think we support localized settings. Which setting did you want to localize?

**Reply by OlziiNaran** [reply]
> I don't think we support localized settings. Which setting did you want to localize?

Exam, TextField，the official language is set according to the country, and other languages are not allowed to be entered.

**Reply by OlziiNaran** [reply]
> I don't think we support localized settings. Which setting did you want to localize?

Or，Code convertion and Saved to DB!

**Comment by jptissot** [comment]
You can also use the [localize] shortcode. You may need to override the template for it to process the shortcode.
I also created a part that allows you to add multilingual content to a single content item [here](https://github.com/StatCan/StatCan.OrchardCore/tree/master/src/Modules/StatCan.OrchardCore.LocalizedText).  Docs: https://statcan.github.io/StatCan.OrchardCore/reference/modules/LocalizedText/
